modules/core/src/scenegraph/gltf-instantiator/gltf-animator.js

```diff
@@ -0,0 +1,95 @@
+import {Matrix4} from 'math.gl';
+
+// TODO: import from loaders.gl?
+export const ATTRIBUTE_TYPE_TO_COMPONENTS = {
+  SCALAR: 1,
+  VEC2: 2,
+  VEC3: 3,
+  VEC4: 4,
+  MAT2: 4,
+  MAT3: 9,
+  MAT4: 16
+};
+
+export const ATTRIBUTE_COMPONENT_TYPE_TO_ARRAY = {
+  5120: Int8Array,
+  5121: Uint8Array,
+  5122: Int16Array,
+  5123: Uint16Array,
+  5125: Uint32Array,
+  5126: Float32Array
+};
+//
+
+
+// TODO: save and do not regenerate
+function accessorToJsArray(accessor) {
+  const ArrayType = ATTRIBUTE_COMPONENT_TYPE_TO_ARRAY[accessor.componentType];
+  const components = ATTRIBUTE_TYPE_TO_COMPONENTS[accessor.type];
+  const length = components * accessor.count;
+  const {buffer, byteOffset} = accessor.bufferView.data;
+
+  return {array: new ArrayType(
+    buffer,
+    byteOffset + (accessor.byteOffset || 0),
+    length
+  ), components};
+}
+
+// TODO: share with GLTFInstantiator
+function applyTRS(gltfNode, node) {
+  node.matrix.identity();
+
+  if (gltfNode.translation) {
+    node.matrix.translate(gltfNode.translation);
+  }
+
+  if (gltfNode.rotation) {
+    const rotationMatrix = new Matrix4().fromQuaternion(gltfNode.rotation);
+    node.matrix.multiplyRight(rotationMatrix);
+  }
+
+  if (gltfNode.scale) {
+    node.matrix.scale(gltfNode.scale);
+  }
+}
+
+export default class GLTFAnimator {
+  constructor(gltf) {
+    this.gltf = gltf;
+    this.animations = this.gltf.animations.map((animation, index) => {
+      const name = animation.name || `Animation-${index}`;
+      const samplers = animation.samplers.map(({input, interpolation = "LINEAR", output}) => ({
+        input: accessorToJsArray(this.gltf.accessors[input]),
+        interpolation,
+        output: accessorToJsArray(this.gltf.accessors[output])
+      }));
+      const channels = animation.channels.map(({sampler, target}) => ({
+        sampler: samplers[sampler],
+        target: this.gltf.nodes[target.node],
+        path: target.path
+      }));
+      return {name, channels, startTime: 0, playing: true};
+    });
+  }
+
+  animate(timeMs) {
+    this.animations.filter(animation => animation.playing).forEach(animation => {
+      animation.absTime = timeMs / 1000;
+      const time = animation.absTime - animation.startTime;
+
+      animation.channels.forEach(({sampler: {input, output}, target, path}) => {
```

Comment:
Can we document this just a bit more generously? I am not super familiar with the glTF animation system. The keyframes are stored in (small) accessors that need to be handled on the CPU?

So every renderframe we need to hook the animation loop to have nodes enable the animator to update their matrix uniforms?

Reply:
I think so. I cannot think of a way to implement this in the shader because shader is not aware of parent nodes.